test/driver.cpp

```diff
@@ -163,3 +172,12 @@ void processCmdLine(const std::vector<string_view>& arguments)
    if (printSummary)
       summarizeTests();
 }
+
+//Replace all instances of a substring with a different substring
+void replace_all(string& str, const string_view substr, 
+   const string& toReplace)
+{
+   string::size_type pos = str.find(substr);
+   for (; pos != string::npos; pos = str.find(substr))
+      str.replace(pos, 4, toReplace);
+}
```

Comment:
Thanks for writing the new function. I agree the last parameter needs to be received by reference.

I see the following issues:
- The function needs to work for any substr, new substr values: it currently does not due to substr length being hardcoded
- The function should work in any context, not just for the limited context of our code. Thus, it is better to be uniform about using either string_view or string for the substr parameters.
- I have some guidelines for when to use const string& parameter vs const string_view&. I am yet to publish that post, but in this case const string& would work just fine. (A function template is actually ideal)
- L181: `pos = str.find(substr)` unnecessarily searches from the beginning of the string. The search can safely start from pos+substr_size, but that would ignore the possibility the new substr can be shorter. And, to ensure termination, we can reasonably assume that the new substr does not contain the replaced substr. We should add an assertion to that effect.

I have the following suggested rewrite of the function. An implementation with unit tests is here: https://godbolt.org/z/7ipMLX. 

Please verify my implementation and tests and let me know if there are any issues. Note that I have intentionally commented out the last test case which causes assertion failure.

```suggestion
//replace all instances of a substring with a different substring
//the new substring should not contain the replaced substring
void replace_all(std::string& str, const std::string& substr, const std::string& new_substr)
{
   assert(new_substr.find(substr) == std::string::npos);

   auto substr_size = substr.size(), new_substr_size = new_substr.size();
   auto pos = str.find(substr);
   for (; pos != std::string::npos; pos = str.find(substr, pos+new_substr_size))
      str.replace(pos, substr_size, new_substr);
}
```

Reply:
I thought a bit more about `replace_all` and concluded that the assertion is unnecessary. I also concluded that the offset used in repeated find could be substr_size in some cases, and new_substr_size in other cases. Also, there is the case where offsetting by substr_size can cause non-termination. (Interestingly, boost library's `replace_all` appears to offset by substr_size).

With these factors in mind, and with some optimization, I revised the implementation and the unit tests: https://godbolt.org/z/4VkDvU

**Please review the revised version and opine.**

Reply:
I don't mean to spam, but I realized that the `replace_all` function was missing a choice of offset to use for "next find". 
For comparison, I also wrote a program using `boost::replace_all` to verify that the results are "correct". Here is the CE link: https://godbolt.org/z/cAM8i8

At this point, I **like some help testing the function** and comparing the results with boost. This mostlu entails constructing test cases. Note that the `replace_all` I wrote lets the programmer choose how and how much to skip over at each find operation. Thus, the result isn't incorrect if it differs from boost, but the difference only means the value passed to parameter `offset_by` isn't the same as what boost implicitly uses. In other words, the test cases should be such that they reveal which value of `offset_by` matches boost, and which value causes deviation from boost. **The outcome of the study will be choose the most appropriate default value for parameter `offset_by`.**

**PS:** It can also be helpful to compare the results with Java's `replace` method for strings: https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/lang/String.html#replace(java.lang.CharSequence,java.lang.CharSequence)

Reply:
@smurthys I agree the next `substr` find should not start at the beginning of the string every time--not just for efficiency's sake--but if `substr` is a substring of  `new_substr`, the function would be caught in an endless loop.

For example, replacing all occurrences of "John" with "Johnathan". If the find begins at the beginning of the string every time, it would find the "John" in "Johnathan" and would change it to "Johnathanathan", with "athan" endlessly being added.

I also agree that the function should work in any context.

I have reviewed your revised implementation and believe that this is a much healthier version of `replace_all`, however, I'm having trouble envisioning a scenario in which a user would want to offset by `substr.size()` rather than `new_substr.size()`. Would you not always want to offset by the size of the new substring in order to avoid altering the text in the new substring?

I will work on creating a comprehensive test driver for `replace_all` that compares the results to `boost::replace_all` and plan to have it done by this Friday.

Thank you for your suggestions and input.

Reply:
The choice of the offset depends on the app. I agree that offset by `new_substr` size is a good choice. Please make that change.

Reply:
I just made the change to offset next find by `new_substr` size.

Reply:
Looks good!